### ISSUE TITLE
Use pairwise summation for floating point sum to reduce round-off error

### ISSUE THREAD

**Issue by Yuhta** [facebookincubator/velox#5504]
### Description

Naive summation of floating point numbers is subject to large round-off error, $O(\epsilon \sqrt{N})$ in average case and $O(\epsilon N)$  in worst case, where $\epsilon$ is machine precision and $N$ is number of summands.

NumPy uses pairwise summation to improve the accuracy ($O(\epsilon \sqrt{\text{log} N})$ in average case and $O(\epsilon \text{log} N)$ in worst case) while keep the performance; we can specialize the sum aggregate function for floating point numbers to do the same.

**Comment by thefloofe22** [comment]
Is this up for grabs?

**Comment by kstiel** [comment]
I want to take a look at this.

**Comment by Glorycs29** [comment]
NumPy and C++ have **similar time and space complexities for pairwise summation**, while Python has slightly higher overhead due to the new list creation in each recursion. However, the overall performance can also depend on other factors such as the _efficiency of the underlying implementation, hardware, and optimizations._

If this issue is not assigned to anyone could you please assign it to me?

**Comment by Yuhta** [comment]
Yes it is up for grabs.  Can you guys negotiate to decide who is going to work on this?

**Comment by Glorycs29** [comment]
Depends on to whom it is assigned.😁

**Comment by Yuhta** [comment]
I would start from the first guy ask.  @thefloofe22 Do you want to take this?

**Comment by thefloofe22** [comment]
Yep, would love to handle this. thanks so much @Yuhta

**Comment by kevinmingtarja** [comment]
Hi, is this still being worked on? If not I'd like to take this on!
